File: pyVersion/sudokuSolver.py

import sys
import time
import copy
from puzzle import *


def solvePuzzle(puzzle):
    """Solve the sudoku puzzle."""

    queue = [ copy.deepcopy( puzzle ) ]

    for p in queue:

        puzzleChanged = True

        while (puzzleChanged):

            puzzleChanged = False

            for s in p.getUnmarked():

                check1 = p.columnCheck( *s )
                check2 = p.rowCheck( *s )
                check3 = p.blockCheck( *s )
                check4 = p.markable( *s )

                puzzleChanged = any( [check1, check2, check3, check4] )

        unmarked = p.getUnmarked()

        if ( len(unmarked) == 0 ):
            print "here1"
            return p

        else:
            s = unmarked.pop()
            possible = p.grid[s].getPossibilities()

            for m in possible:
                pGuess = copy.deepcopy( p )
                pGuess.markSquare( s[0], s[1], m )

                if ( len(pGuess.getUnmarked()) == 0 ):
                    print "here2"
                    return pGuess
                queue.append( pGuess )

    print "here3"
    return queue[0]


if __name__ == "__main__":

    path = sys.argv[1]

    puzzle = Puzzle.fromFile( path )

    print "Puzzle to solve:"
    puzzle.printPuzzle()

    t1 = time.time()
    solvedPuzzle = solvePuzzle( puzzle )
    t2 = time.time()

    print "Solved puzzle:"
    solvedPuzzle.printPuzzle()

    print "solved in {t}s\n".format(t = t2 - t1)

File: cppVersion/sudokuSolver.cpp

#include <queue>
#include <vector>
#include <set>
#include <utility>
#include <iostream>
#include "puzzle.hpp"


Puzzle solvePuzzle( Puzzle puzzle ) {
   /*Solve the sudoku puzzle.*/

   std::queue<Puzzle> queue;
   queue.push( puzzle );

   while ( not queue.empty() ) {

      Puzzle p = queue.front();
      queue.pop();

      bool puzzleChanged = true;

      while (puzzleChanged) {

         puzzleChanged = false;

         for ( auto &s : p.getUnmarked() ) {

            puzzleChanged |= p.columnCheck( s.first, s.second );
            puzzleChanged |= p.rowCheck( s.first, s.second );
            puzzleChanged |= p.blockCheck( s.first, s.second );
            puzzleChanged |= p.markable( s.first, s.second );
         }
      }

      std::vector< std::pair<int,int> > *unmarked = p.getUnmarkedRef();

      if ( (*unmarked).size() == 0 ) {
         return p;
      }
      else {
         std::pair<int,int> s = (*unmarked).front();
         (*unmarked).erase( (*unmarked).begin() );
         std::set<int> possible = p.getGrid().at(s).getPossibilities();

         for (auto &m : possible) {
            Puzzle pGuess(p);
            pGuess.markSquare( s.first, s.second, m );

            if ( pGuess.getUnmarked().size() == 0 ) {
               return pGuess;
            }

            queue.push( pGuess );
         }
      }
   }

   return queue.front();
}


int main ( int argc, char* argv[] ) {

   Puzzle puzzle( argv[1] );

   std::cout << "\nPuzzle to solve:" <<std::endl;
   puzzle.printPuzzle();

   Puzzle solvedPuzzle = solvePuzzle( puzzle );

   std::cout << "Solved puzzle:" <<std::endl;
   solvedPuzzle.printPuzzle();

   return 0;
}

